vehicle_v1:

  # Publish all joint states -----------------------------------
  joint_state_controller:
    type: joint_state_controller/JointStateController
    publish_rate: 50  
  # gazebo_ros_control:
  #   pid_gains:
  #     front_rev_joint:
  #       p: 100.0
  #       i: 0.01
  #       d: 10.0
  #     back_cont_joint:
  #       p: 1.5
  #       i: 1.0
  #       d: 0.0
  #       i_clamp: 10.0
  # gazebo_ros_control/pid_gains:
  #   front_rev_joint: {p: 20.0, i: 0.01, d: 8.0, i_clamp: 10.0}
  #   # front_cont_joint: {p: 1.5, i: 1.0, d: 0.0, i_clamp: 10.0}
  #   # back_rev_joint: {p: 100.0, i: 0.01, d: 10.0}
  #   back_cont_joint: {p: 1.5, i: 1.0, d: 0.0, i_clamp: 10.0}
    
  # Position Controllers ---------------------------------------
  front_left_rev_controller:
    type: effort_controllers/JointPositionController
    joint: front_left_rev_joint
    pid: {p: 10.0, i: 0.0, d: 0.5}

  front_right_rev_controller:
    type: effort_controllers/JointPositionController
    joint: front_right_rev_joint
    pid: {p: 10.0, i: 0.0, d: 0.5}

  front_left_cont_controller:
    type: effort_controllers/JointVelocityController
    joint: front_left_cont_joint
    pid: {p: 1.5, i: 1.0, d: 0.0, i_clamp: 10.0}

  front_right_cont_controller:
    type: effort_controllers/JointVelocityController
    joint: front_right_cont_joint
    pid: {p: 1.5, i: 1.0, d: 0.0, i_clamp: 10.0}

  # mobile_base_controller:
  #   type        : "ackermann_steering_controller/AckermannSteeringController"
  #   rear_wheel: 'back_cont_joint'
  #   front_steer: 'front_rev_joint'
  #   publish_rate: 50.0               # default: 50
  #   pose_covariance_diagonal : [0.001, 0.001, 1000000.0, 1000000.0, 1000000.0, 1000.0]
  #   twist_covariance_diagonal: [0.001, 0.001, 1000000.0, 1000000.0, 1000000.0, 1000.0]

  #   # Wheel separation between the rear and the front, and diameter of the rear. 
  #   # These are both optional.
  #   # ackermann_steering_controller will attempt to read either one or both from the
  #   # URDF if not specified as a parameter.
  #   wheel_separation_h : 0.1
  #   wheel_radius : 0.025

  #   # Wheel separation and radius multipliers for odometry calibration.
  #   # wheel_separation_h_multiplier: 1.0 # default: 1.0
  #   # wheel_radius_multiplier    : 1.0 # default: 1.0

  #   # # Steer position angle multipliers for fine tuning.
  #   # steer_pos_multiplier       : 1.0

  #   # Velocity commands timeout [s], default 0.5
  #   cmd_vel_timeout: 50

  #   # Base frame_id
  #   base_frame_id: base_link #default: base_link

  #   # Odom frame_id
  #   odom_frame_id: odom

  #   # Velocity and acceleration limits
  #   # Whenever a min_* is unspecified, default to -max_*
  #   # linear:
  #   #   x:
  #   #     has_velocity_limits    : true
  #   #     max_velocity           : 1.0  # m/s
  #   #     min_velocity           : -0.5 # m/s
  #   #     has_acceleration_limits: true
  #   #     max_acceleration       : 0.8  # m/s^2
  #   #     min_acceleration       : -0.4 # m/s^2
  #   #     has_jerk_limits        : true
  #   #     max_jerk               : 5.0 # m/s^3

  #   # angular:
  #   #   z:
  #   #     has_velocity_limits    : true
  #   #     max_velocity           : 1.7  # rad/s
  #   #     has_acceleration_limits: true
  #   #     max_acceleration       : 1.5  # rad/s^2
  #   #     has_jerk_limits        : true
  #   #     max_jerk               : 2.5 # rad/s^3